### ISSUE TITLE
about the implementation of scaled attention

### ISSUE THREAD

**Issue by yiyixuxu** [lucidrains/denoising-diffusion-pytorch#78]
Hi Lucidrans:

Thanks for your awesome work. I'm working on a flax version of ddpm based your implementation. I found the implementation of Attention is a little bit different: you first normalized the `q`, `k` vector and then scaled it up with a fixed number `16` ( instead scale down with  `dim_head ** -0.5`)  - is this intended? 

Thanks!
YiYi

```python
class Attention(nn.Module):
    def __init__(self, dim, heads = 4, dim_head = 32, scale = 16):
        super().__init__()
        self.scale = scale
        self.heads = heads
        hidden_dim = dim_head * heads
        self.to_qkv = nn.Conv2d(dim, hidden_dim * 3, 1, bias = False)
        self.to_out = nn.Conv2d(hidden_dim, dim, 1)

    def forward(self, x):
        b, c, h, w = x.shape
        qkv = self.to_qkv(x).chunk(3, dim = 1)
        q, k, v = map(lambda t: rearrange(t, 'b (h c) x y -> b h c (x y)', h = self.heads), qkv)

        q, k = map(l2norm, (q, k))

        sim = einsum('b h d i, b h d j -> b h i j', q, k) * self.scale
        attn = sim.softmax(dim = -1)

        out = einsum('b h i j, b h d j -> b h i d', attn, v)
        out = rearrange(out, 'b h (x y) d -> b (h d) x y', x = h, y = w)
        return self.to_out(out)
```

**Comment by lucidrains** [comment]
@yiyixuxu Hi Yiyi! Yes, very observant! This is actually a [variant](https://arxiv.org/abs/2010.04245) of normal attention that eliminates a common problem in attention / transformers; overflow when calculating the similarities between queries and keys

I am planning on taking a risk and adopting it for all my future work (it is not widely used yet, but in my experiments, not any worse than normal attention)

**Comment by lucidrains** [comment]
@yiyixuxu Hi Yiyi! Yes, very observant! This is actually a [variant](https://arxiv.org/abs/2010.04245) of normal attention that eliminates a common problem in attention / transformers; overflow when calculating the similarities between queries and keys

I am planning on taking a risk and adopting it for all my future work (it is not widely used yet, but in my experiments, not any worse than normal attention)

**Comment by yiyixuxu** [comment]
@lucidrains Thank you for explaining!
